### PR TITLE
Enhance EncryptAlgorithm and QueryAssistedEncryptAlgorithm

### LINKED ISSUE #11689: [enhancement]how about add context parameters on encrypt and decrypt method
## Feature Request

**For English only**, other languages will not accept.

### Describe the feature you would like.
We need different key on each table/column to encrypt and decrypt with, how about add content parameters as below code, for example: logicTable, logicColumn etc? 

`/**
 * Encryptor.
 */
public interface Encryptor extends TypeBasedSPI {
    
    /**
     * Initialize.
     */
    void init();
    
    /**
     * Encode.
     * 
     * @param plaintext plaintext
     * @param encryptContextMap encrypt context map(logicTable and logicColumn)
     * @return ciphertext
     */
    String encrypt(Object plaintext, **Map<String, Object> encryptContextMap**);
    
    /**
     * Decode.
     * 
     * @param ciphertext ciphertext
     * @param decryptContextMap decrypt context map(logicTable and logicColumn)
     * @return plaintext
     */
    Object decrypt(String ciphertext, **Map<String, Object> decryptContextMap**);
}`

### PR DESCRIPTION
**Pull request by cheese8**
@tristaZero Please help to review, thanks.

Fixes #11689. 

Changes proposed in this pull request:

add SchemaAvailable to collect schemaName from dml StatementContext;
modify directly on EncryptAlgorithm & QueryAssistedEncryptAlgorithm, and add parameter Map<String, String> encryptContextMap;
migrate RuleConfigurationLogger to log and debug the value set on EncryptConfiguration, will refactoring it later;

### CODE REVIEW

**Comment by cheese8** [comment]
Please review on reimplementation with SchemaAvailable which different from previous hardcoding way, and rePR this ticket, sorry for inconvenience.

**Review by strongduanmu — CHANGES_REQUESTED** [review]
@cheese8 Can you fix ci error first?

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-api/src/main/java/org/apache/shardingsphere/encrypt/spi/EncryptAlgorithm.java [hunk 1]
```diff
@@ -28,10 +30,11 @@
     /**
      * Encode.
      *
+     * @param encryptContextMap encrypt context map which includes schema, owner, table and column keys.
      * @param plaintext plaintext
      * @return ciphertext
      */
-    String encrypt(Object plaintext);
+    String encrypt(Object plaintext, Map<String, String> encryptContextMap);
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-api/src/main/java/org/apache/shardingsphere/encrypt/spi/EncryptAlgorithm.java:37]: @cheese8 Why only encrypt adds parameters, but decrypt remains the same?

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-api/src/main/java/org/apache/shardingsphere/encrypt/spi/EncryptAlgorithm.java [hunk 2]
```diff
@@ -28,10 +30,11 @@
     /**
      * Encode.
      *
+     * @param encryptContextMap encrypt context map which includes schema, owner, table and column keys.
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-api/src/main/java/org/apache/shardingsphere/encrypt/spi/EncryptAlgorithm.java:33]: @cheese8 Can we change this param type to Properties?

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rewrite/parameter/impl/EncryptInsertOnDuplicateKeyUpdateValueParameterRewriter.java [hunk 3]
```diff
@@ -55,15 +59,16 @@ public void rewrite(final ParameterBuilder parameterBuilder, final InsertStateme
             int columnIndex = index;
             String encryptLogicColumnName = onDuplicateKeyUpdateValueContext.getColumn(columnIndex).getIdentifier().getValue();
             Optional<EncryptAlgorithm> encryptorOptional = getEncryptRule().findEncryptor(tableName, encryptLogicColumnName);
+            Map<String, String> encryptContextMap = new EncryptContext(insertStatementContext.getSchemaName(), tableName, encryptLogicColumnName).of();
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rewrite/parameter/impl/EncryptInsertOnDuplicateKeyUpdateValueParameterRewriter.java:62]: @cheese8 Can we consider that the logic of encryption and decryption parameters is implemented by the user in the form of SPI?

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rule/EncryptContext.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    
+    private String column;
+    
+    public EncryptContext(String schema, String table, String column) {
+        this.schema = schema;
+        this.table = table;
+        this.column = column;
+    }
+    
+    public Map<String, String> of(){
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rule/EncryptContext.java:48]: @cheese8 Can we change this result type to Properties?

File: shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rule/EncryptContext.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+/**
+ * Encrypt context.
+ */
+@Getter
+@Setter
+public final class EncryptContext {
+    
+    private String schema;
+    
+    private String owner = "";
```
Comment (strongduanmu) [inline shardingsphere-features/shardingsphere-encrypt/shardingsphere-encrypt-core/src/main/java/org/apache/shardingsphere/encrypt/rule/EncryptContext.java:36]: @cheese8 Please remove this default value and add final for this field.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/log/RuleConfigurationLogger.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Rule configuration logger.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@Slf4j
+public final class RuleConfigurationLogger {
```
Comment (strongduanmu) [inline shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/log/RuleConfigurationLogger.java:34]: @cheese8 Can we remove this class?

**Review by strongduanmu — CHANGES_REQUESTED** [review]

**Pull request closed without merging**